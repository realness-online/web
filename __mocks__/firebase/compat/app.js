export default {
  initializeApp: vi.fn(),
  apps: []
}
