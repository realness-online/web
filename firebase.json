{
  "storage": {
    "rules": "storage.rules"
  },
  "hosting": {
    "public": "dist",
    "ignore": [
      "firebase.json",
      "**/.*",
      "**/node_modules/**"
    ],
    "rewrites": [
      {
        "source": "**",
        "destination": "/index.html"
      }
    ],
    "headers": [
      {
        "source": "**",
        "headers": [
          {
            "key": "Content-Security-Policy",
            "value": "script-src 'unsafe-eval' 'unsafe-inline' 'self' https://www.google.com https://apis.google.com https://www.gstatic.com https://storage.googleapis.com/"
          }
        ]
      },
      {
        "source": "**",
        "headers": [
          {
            "key": "Cache-Control",
            "value": "max-age=31536000"
          }
        ]
      },
      {
        "source": "/service.worker.js",
        "headers": [
          {
            "key": "Cache-Control",
            "value": "no-cache"
          }
        ]
      }
    ]
  }
}
