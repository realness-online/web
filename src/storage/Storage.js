// https://developers.caffeina.com/object-composition-patterns-in-javascript-4853898bb9d0
import Item from '@/modules/item'
import Paged from '@/storage/Paged'
import Large from '@/storage/Large'
import Cloud from '@/storage/Cloud'

class Storage {
  constructor(type,
              selector = `[itemprop="/${type}"]`,
              filename = `${type}/index.html`,
              content_type = 'text/html') {
    this.type = type
    this.selector = selector
    this.filename = filename
    this.metadata = { 'contentType': content_type }
  }
  as_list() {
    return Item.get_items(localStorage.getItem(this.selector))
  }
  as_object() {
    return Item.get_first_item(localStorage.getItem(this.selector))
  }
  async save(items = document.querySelector(this.selector)) {
    if (!items) return
    localStorage.setItem(this.selector, items.outerHTML)
  }
}
class Person extends Cloud(Storage) {
  constructor() {
    super('person', '[itemid="person.html"]', 'person.html')
  }
}
class Post extends Paged(Cloud(Storage)) {
  constructor() {
    super('posts')
  }
}
class SVG extends Large(Cloud(Storage)) {}

export default Storage
export const relations_storage = new Storage('relations')
export const person_storage = new Person()
export const posts_storage = new Posts()
export const avatars_storage = new SVG('avatars')
export const posters_storage = new SVG('posters')
