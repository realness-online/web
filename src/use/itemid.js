// {base_url}/{:author}/{:type}/{:created_at}
import get_item from '@/use/item'
import { does_not_exist } from '@/use/sync'
import { current_user, url, directory } from '@/use/serverless'
import { get, set, keys } from 'idb-keyval'
class Directory {
  constructor() {
    this.items = []
    this.types = [] // folders are types in our vocabulary
  }
}
// Expensive to call
export async function load(itemid, me = localStorage.me) {
  let item
  if (~itemid.indexOf(me)) {
    item = localStorage.getItem(itemid)
    if (item) return get_item(item)
    else if (as_type(itemid) === 'relations') return []
  }
  const result = await get(itemid)
  item = get_item(result)
  if (item) return item
  item = await load_from_network(itemid, me)
  if (item) return item
  return null
}
export async function list(itemid, me = localStorage.me) {
  try {
    const item = await load(itemid, me)
    if (item) return type_as_list(item)
    else return []
  } catch {
    return []
  }
}
export async function load_from_network(itemid, me = localStorage.me) {
  const url = await as_download_url(itemid, me)
  if (url) {
    console.info('download', itemid)
    const server_text = await (await fetch(url)).text()
    await set(itemid, server_text)
    return get_item(server_text)
  } else return null
}
export async function load_directory_from_network(itemid) {
  if (navigator.onLine && current_user.value) {
    const path = as_directory_id(itemid)
    const meta = new Directory()
    console.info('request:directory', itemid)
    const folder = await directory(`people/${path}`)
    folder.items.forEach(item => meta.items.push(item.name.split('.')[0]))
    folder.prefixes.forEach(prefix => meta.types.push(prefix.name))
    await set(path, meta)
    return meta
  }
}
export async function as_directory(itemid) {
  const path = as_directory_id(itemid)
  const cached = await get(path)
  if (cached) return cached
  let directory = await build_local_directory(itemid)
  if (navigator.onLine && current_user.value)
    directory = await load_directory_from_network(itemid)
  return directory
}
export async function as_download_url(itemid) {
  if (!current_user.value) return null
  try {
    console.info('request:location', itemid)
    return await url(as_filename(itemid))
  } catch (e) {
    if (e.code === 'storage/object-not-found') {
      console.warn(itemid, '=>', as_filename(itemid))
      const index = (await get('sync:index')) || {}
      index[itemid] = does_not_exist
      await set('sync:index', index)
      return null
    } else throw e
  }
}

// Cheap to call
const created_at = ['posters', 'animations']
const has_history = ['statements', 'events']
export function is_history(itemid) {
  const parts = as_path_parts(itemid)
  if (has_history.includes(as_type(itemid)) && parts.length === 3) return true
  else return false
}
export function as_filename(itemid) {
  let filename = itemid
  if (itemid.startsWith('/+')) filename = `/people${filename}`
  if (created_at.includes(as_type(itemid)) || is_history(itemid))
    return `${filename}.html`
  else return `${filename}/index.html`
}
export function as_storage_path(itemid) {
  const path_parts = as_path_parts(itemid)
  let path = `/${path_parts[0]}`
  if (itemid.startsWith('/+')) path = `/people${path}`
  switch (path_parts.length) {
    case 0:
      return null
    case 1:
      return path
    default:
      return `${path}/${path_parts[1]}`
  }
}
export async function build_local_directory(itemid) {
  const path = as_directory_id(itemid)
  const directory = new Directory()
  const everything = await keys()
  everything.forEach(itemid => {
    if (as_directory_id(itemid) === path) {
      const id = as_created_at(itemid)
      if (id) directory.items.push(id)
    }
  })
  return directory
}
export function as_directory_id(itemid) {
  const parts = as_path_parts(itemid)
  return `/${parts[0]}/${parts[1]}/`
}
export function as_path_parts(itemid) {
  const path = itemid.split('/')
  if (path[0].length === 0) path.shift()
  return path
}
export function as_author(itemid) {
  const path = as_path_parts(itemid)
  const author = path[0] || ''
  if (author.startsWith('+1')) return `/${path[0]}`
  else return null
}
export function as_type(itemid) {
  const path = as_path_parts(itemid)
  if (path[1]) return path[1]
  if (itemid.startsWith('/+')) return 'person'
  else return null
}
export function as_created_at(itemid) {
  const path = as_path_parts(itemid)
  if (path.length === 1) return parseInt(path[0])
  return parseInt(path[2])
}
export function as_query_id(itemid) {
  return itemid.substring(2).replace('/', '-').replace('/', '-')
}
export function as_fragment_id(itemid) {
  return `#${as_query_id(itemid)}`
}
export function type_as_list(item) {
  // Returns a list even if loading the item fails
  // the microdata spec requires properties values to
  // single value and iterable
  if (!item) return []
  const list = item[as_type(item.id)]
  if (list && Array.isArray(list)) return list
  else if (list) return [list]
  else return []
}
