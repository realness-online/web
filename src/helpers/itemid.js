// {base_url}/{:author}/{:type}/{:created_at}
import firebase from 'firebase/app'
import 'firebase/storage'
import 'firebase/auth'
import get_item from '@/modules/item'
import { does_not_exist } from '@/persistance/Cloud.sync'
import { get, set } from 'idb-keyval'
// Expensive to call
export async function load(itemid, me = localStorage.me) {
  let item
  if (~itemid.indexOf(me)) {
    item = localStorage.getItem(itemid)
    if (item) return get_item(item)
    else if (as_type(itemid) === 'relations') return []
  }
  const result = await get(itemid)
  item = get_item(result)
  if (item) return item
  item = await load_from_network(itemid, me)
  if (item) return item
  return null
}
export async function list(itemid, me = localStorage.me) {
  try {
    const item = await load(itemid, me)
    if (item) return type_as_list(item)
    else return []
  } catch {
    return []
  }
}
export async function load_from_network(itemid, me = localStorage.me) {
  const url = await as_download_url(itemid, me)
  if (url) {
    console.info('download', itemid)
    const server_text = await (await fetch(url)).text()
    await set(itemid, server_text)
    return get_item(server_text)
  } else return null
}
export async function as_directory(itemid) {
  const path = as_directory_id(itemid)
  const cached = await get(path)
  if (cached) return cached
  else if (navigator.onLine && firebase.auth().currentUser) {
    const meta = { items: [], types: [] } // folders are types in our vocabulary
    console.info('request:directory', itemid)
    const directory = await firebase.storage().ref().child(`people/${path}`).listAll()
    directory.items.forEach(item => meta.items.push(item.name.split('.')[0]))
    directory.prefixes.forEach(prefix => meta.types.push(prefix.name))
    await set(path, meta)
    return meta
  } else return null
}
export async function as_download_url(itemid) {
  if (!firebase.auth().currentUser) return null
  const storage = firebase.storage().ref()
  try {
    console.info('request:location', itemid)
    return await storage.child(as_filename(itemid)).getDownloadURL()
  } catch (e) {
    if (e.code === 'storage/object-not-found') {
      console.warn(itemid, '=>', as_filename(itemid))
      const index = (await get('sync:index')) || {}
      index[itemid] = does_not_exist
      await set('sync:index', index)
      return null
    } else throw e
  }
}

// Cheap to call
const created_at = ['avatars', 'posters', 'animations']
const has_history = ['statements', 'events']
export function is_history(itemid) {
  const parts = as_path_parts(itemid)
  if (has_history.includes(as_type(itemid)) && parts.length === 3) return true
  return false
}
export function as_filename(itemid) {
  let filename = itemid
  if (itemid.startsWith('/+')) filename = `/people${filename}`
  if (created_at.includes(as_type(itemid)) || is_history(itemid)) return `${filename}.html`
  else return `${filename}/index.html`
}
export function as_storage_path(itemid) {
  const path_parts = as_path_parts(itemid)
  let path = `/${path_parts[0]}`
  if (itemid.startsWith('/+')) path = `/people${path}`
  switch (path_parts.length) {
    case 0:
      return null
    case 1:
      return path
    default:
      return `${path}/${path_parts[1]}`
  }
}
export function as_directory_id(itemid) {
  const parts = as_path_parts(itemid)
  return `/${parts[0]}/${parts[1]}/`
}
export function as_path_parts(itemid) {
  const path = itemid.split('/')
  if (path[0].length === 0) path.shift()
  return path
}
export function as_author(itemid) {
  const path = as_path_parts(itemid)
  const author = path[0] || ''
  if (author.startsWith('+1')) return `/${path[0]}`
  else return null
}
export function as_type(itemid) {
  const path = as_path_parts(itemid)
  if (path[1]) return path[1]
  if (itemid.startsWith('/+')) return 'person'
  else return null
}
export function as_created_at(itemid) {
  const path = as_path_parts(itemid)
  return parseInt(path[2])
}
export function as_query_id(itemid) {
  return itemid.substring(2).replace('/', '-').replace('/', '-')
}
export function as_fragment(itemid) {
  return `#${as_query_id(itemid)}`
}
export function type_as_list(item) {
  // Returns a list even if loading the item fails
  // the microdata spec requires properties values to
  // single value and iterable
  if (!item) return []
  const list = item[as_type(item.id)]
  if (list && Array.isArray(list)) return list
  else if (list) return [list]
  else return []
}
