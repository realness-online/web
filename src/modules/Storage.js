import Item from '@/modules/Item'
import * as firebase from 'firebase/app'
import 'firebase/auth'
import 'firebase/storage'

class Storage {
  static hydrate(item_as_string) {
    return document.createRange().createContextualFragment(item_as_string)
  }
  constructor(item_type, selector = `[itemtype="/${item_type}"]`,
    filename = `${item_type}.html`) {
    this.item_type = item_type
    this.selector = selector
    this.filename = filename
    this.metadata = {'contentType': 'text/html'}
  }
  from_storage() {
    let storage_string = localStorage.getItem(this.item_type)
    return Storage.hydrate(storage_string)
  }
  as_list() {
    return Item.get_items(this.from_storage())
  }
  as_object() {
    return Item.get_first_item(this.from_storage())
  }
  save() {
    // save the information to local storage and if appropriate to the server
    let items = document.querySelector(this.selector)
    if (!items) { return false }
    items = items.outerHTML
    localStorage.setItem(this.item_type, items)
    if (['person', 'posts'].includes(this.item_type)) {
      this.persist(items)
    }
    return true
  }
  sync() {
    // check if an 'item_type'.html already exist for this user.
    // If so merge it's content whith what's stored locally
    const url = this.get_server_file_url()
    if (['person', 'posts'].includes(this.item_type) && navigator.onLine) {
      firebase.auth().onAuthStateChanged(user => {
        if (user) {
          const doc_u_path = `/people/${user.phoneNumber}/${this.filename}`
          firebase.storage().ref().child(doc_u_path).getDownloadURL().then((url) => {
            fetch(url).then(response => {
              response.text().then(server_text => {
                const server_as_fragment = Storage.hydrate(server_text)
                let server_as_list = Item.get_items(server_as_fragment)
                console.log(server_as_list)
              })
              let items = [...this.from_storage, ...from_server]
              // console.log(items)
            })
          })
        }
      })
    }
  }

  get_server_file_url() {

  }

  persist(doc_u_ment, doc_u_path) {
    firebase.auth().onAuthStateChanged(user => {
      if (user && navigator.onLine) {
        const file = new File([doc_u_ment], this.filename)
        if (!doc_u_path) {
          doc_u_path = `/people/${user.phoneNumber}/${this.filename}`
        }
        firebase.storage().ref()
          .child(doc_u_path)
          .put(file, this.metadata)
          .catch(console.log.bind(console))
      }
    })
  }
}

export default Storage
export const person_storage = new Storage('person')
export const posts_storage = new Storage('posts', '[itemprop=posts]')
export const activity_storage = new Storage('activity', '[itemprop=activity]')
export const relations = new Storage('relations', '[itemprop=relations]')
export const phonebook = new Storage('phonebook')

//  use this for firebase logging
// .then( snapshot => {
//   console.log('Uploaded', snapshot.totalBytes, 'bytes.')
//   console.log(snapshot.metadata)
//   console.log(snapshot.fullPath)
//   console.log('File available at', snapshot.downloadURL)
// })
