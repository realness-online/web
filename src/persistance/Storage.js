/** @typedef {import('@/types').Item_Id} Item_Id */

// https://developers.caffeina.com/object-composition-patterns-in-javascript-4853898bb9d0
import Local from '@/persistance/Local'
import Large from '@/persistance/Large'
import Cloud from '@/persistance/Cloud'
import Paged from '@/persistance/Paged'
import { current_user, upload } from '@/use/serverless'
import { as_type, as_filename } from '@/utils/itemid'
import { get } from 'idb-keyval'
import { prepare_upload_html } from '@/utils/upload_processor'

export default class Storage {
  /**
   * @param {Item_Id} itemid
   */
  constructor(itemid) {
    this.id = itemid
    this.type = as_type(itemid)
    this.metadata = { contentType: 'text/html' }
  }
}

/** @extends {Storage} */
export class Poster extends Large(Cloud(Storage)) {}

/** @extends {Storage} */
export class Me extends Cloud(Local(Storage)) {
  constructor() {
    super(localStorage.me)
  }
}

/** @extends {Storage} */
export class Relation extends Local(Storage) {
  constructor() {
    super(`${localStorage.me}/relations`)
  }
}

/** @extends {Storage} */
export class Statement extends Paged(Cloud(Local(Storage))) {
  constructor() {
    super(`${localStorage.me}/statements`)
  }
}

/** @extends {Storage} */
export class Event extends Paged(Cloud(Local(Storage))) {
  constructor() {
    super(`${localStorage.me}/events`)
  }
}

/** @extends {Storage} */
export class Offline extends Cloud(Storage) {
  async save() {
    const outerHTML = await get(this.id)
    if (outerHTML) await super.save({ outerHTML })
  }
}

/** @extends {Storage} */
export class History extends Cloud(Storage) {
  /**
   * @param {Item_Id} itemid
   */
  constructor(itemid) {
    super(itemid)
  }

  async save(items) {
    // on purpose doesn't call super.save
    if (!items) return false
    if (current_user.value && navigator.onLine) {
      const { compressed, metadata } = await prepare_upload_html(items)
      try {
        await upload(as_filename(this.id), compressed, metadata)
      } catch (e) {
        console.error(e)
        return false
      }
      return true
    }
    return false
  }
}
