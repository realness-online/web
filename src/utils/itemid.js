/** @typedef {import('@/types').Id} Id */
/** @typedef {import('@/types').Type} Type */
/** @typedef {import('@/types').Created} Created */
/** @typedef {import('@/types').Author} Author */
/** @typedef {import('@/types').Item} Item */
import { get, set } from 'idb-keyval'
import get_item from '@/utils/item'
import { DOES_NOT_EXIST } from '@/use/sync'
import { url } from '@/utils/serverless'
import { decompress_html } from '@/utils/upload-processor'
import { as_directory_id } from '@/persistance/Directory'
import { newest_timestamp_first } from '@/utils/sorting'
/**
 * @type {Type[]}
 */
const has_archive = ['posters']

/**
 * @type {Type[]}
 */
const has_history = ['statements', 'events']

// this leaves me, history, relations, and offline

/**
 * @param {Id} itemid
 * @returns {Promise<string>}
 */
export const as_filename = async itemid => {
  let filename = itemid
  if (itemid.startsWith('/+')) filename = `people${itemid}`

  if (has_archive.includes(as_type(itemid))) {
    const archive = await as_archive(itemid)
    console.log('archive name', archive)
    if (archive) return `${archive}.html.gz`
    return `${filename}.html.gz`
  } else if (is_history(itemid)) return `${filename}.html.gz`

  return `${filename}/index.html.gz`
}

/**
 * @param {Id} itemid
 * @returns {Promise<string | null>}
 */
export const load_from_network = async itemid => {
  const url = await as_download_url(itemid)

  if (url) {
    const response = await fetch(url)

    // Check Content-Encoding header
    const content_encoding = response.headers.get('Content-Encoding')
    const compressed_html = await response.arrayBuffer()
    let html = null
    // If no content encoding or 'identity', data is already decompressed
    if (!content_encoding || content_encoding === 'identity')
      html = new TextDecoder().decode(compressed_html)
    else html = await decompress_html(compressed_html)

    if (!html) return null
    console.log('storage', itemid)
    await set(itemid, html)
    return get_item(html)
  }
  return null
}

/**
 * @param {Id} itemid
 * @param {Author} me
 * @returns {Promise<Item | null>}
 */
export const load = async (itemid, me = localStorage.me) => {
  console.trace('load', itemid)
  let item
  if (~itemid.indexOf(me)) {
    item = localStorage.getItem(itemid)
    if (item) return get_item(item)
    else if (as_type(itemid) === 'relations') return []
  }
  const result = await get(itemid)
  item = get_item(result)
  if (item) return item
  try {
    item = await load_from_network(itemid, me)
  } catch (e) {
    if (e.code === 'storage/unauthorized') return null
    throw e
  }
  if (item) return item
  return null
}

/**
 * @param {Id} itemid
 * @param {Author} me
 * @returns {Promise<Type[]>}
 */
export const list = async (itemid, me = localStorage.me) => {
  try {
    const item = await load(itemid, me)
    if (item) return type_as_list(item)
    return []
  } catch {
    return []
  }
}

/**
 * @param {Id} itemid
 * @returns {Promise<string | null>}
 */
export const as_download_url = async itemid => {
  if (itemid.startsWith('/+/')) return null
  try {
    return await url(await as_filename(itemid))
  } catch (e) {
    if (e.code === 'storage/object-not-found') {
      console.trace(itemid, '=>', await as_filename(itemid))
      const index = (await get('sync:index')) || {}
      index[itemid] = DOES_NOT_EXIST
      await set('sync:index', index)
      return null
    }
    throw e
  }
}

/**
 * Splits an item ID path into its constituent parts
 * @param {Id} itemid - The full item path (e.g. '/+123456/statements/789')
 * @returns {[Author] | [Author, Type] | [Author, Type, Created]} Array of path parts where:
 *   [0] = author ID (e.g. '+123456')
 *   [1] = item type (e.g. 'statements', 'events', 'posters', etc.)
 *   [2] = created_at timestamp (if applicable)
 * @example
 * as_path_parts('/+123456/statements/789') // ['+123456', 'statements', '789']
 * as_path_parts('/+123456') // ['+123456']
 * as_path_parts('/+123456/statements/789/') // ['+123456', 'statements', null, '789']
 */
export const as_path_parts = itemid => {
  const path = itemid.split('/')
  if (path[0].length === 0) path.shift()
  if (itemid.endsWith('/')) {
    //is a directory
    const [author, type, created = null, archive = null] = path
    if (!created && !archive) return [author, type, 'index', null]
    if (created) return [author, type, null, created]
  }
  return path
}

/**
 * @param {Id} itemid
 * @returns {string | null}
 */
export const as_storage_path = itemid => {
  const path_parts = as_path_parts(itemid)
  let path = `/${path_parts[0]}`
  if (itemid.startsWith('/+')) path = `/people${path}`
  switch (path_parts.length) {
    case 0:
      return null
    case 1:
      return path
    default:
      return `${path}/${path_parts[1]}`
  }
}

/**
 * @param {Id} itemid
 * @returns {string | null}
 */
export const as_author = itemid => {
  const path = as_path_parts(itemid)
  const author = path[0] || ''
  if (author.startsWith('+1')) return `/${path[0]}`
  return null
}

/**
 * @param {Id} itemid
 * @returns {Type | null}
 */
export const as_type = itemid => {
  const path = as_path_parts(itemid)
  if (path[1]) return path[1]
  if (itemid.startsWith('/+')) return 'person'
  return null
}

/**
 * @param {Id} itemid
 * @returns {Created | null}
 */
export const as_created_at = itemid => {
  const path = as_path_parts(itemid)
  /** @type {1 | 2 | 3 | 4} */
  const path_length = path.length
  if (path_length === 3) return parseInt(path[2])
  if (path_length === 4) return parseInt(path[3])
  return null
}

/**
 * @param {Id} itemid
 * @returns {string}
 */
export const as_query_id = itemid =>
  itemid.substring(2).replace('/', '-').replace('/', '-')
//

/**
 * @param {Id} itemid
 * @returns {string}
 */
export const as_fragment_id = itemid => `#${as_query_id(itemid)}`

/**
 * @param {Item} item
 * @returns {Type[]}
 */
export const type_as_list = item => {
  // Returns a list even if loading the item fails
  // the microdata spec requires properties values to
  // single value and iterable
  if (!item) return []
  const list = item[as_type(item.id)]
  if (list && Array.isArray(list)) return list
  else if (list) return [list]
  return []
}
/**
 * @param {Id} itemid
 * @returns {boolean}
 */
export const is_history = itemid => {
  const parts = as_path_parts(itemid)
  if (has_history.includes(as_type(itemid)) && parts.length === 3) return true
  return false
}

/**
 * Determines if an item needs to be loaded from an archive
 * @param {Id} itemid - The ID of the item to check
 * @returns {Promise<string | null>} - Returns null if the item exists in the main items list,
 *                                    otherwise returns the archive path if found in an archive,
 *                                    or null if not found in either location
 */
export const as_archive = async itemid => {
  const directory = await get(as_directory_id(itemid))
  if (!directory) return null

  const { items = [], archive = [] } = directory
  const created = as_created_at(itemid)
  if (!created) return null

  if (items.includes(created.toString())) return null

  if (archive.includes(created))
    return `people${as_author(itemid)}/${as_type(itemid)}/${created}/${created}`

  const sorted_archive = [...archive].sort(newest_timestamp_first)

  let closest_timestamp = null // Find the closest archive timestamp that's GREATER than or EQUAL to the created timestamp
  for (const timestamp of sorted_archive)
    if (created <= timestamp) {
      closest_timestamp = timestamp
      break // Stop at the first timestamp that's greater than or equal to created
    }

  if (!closest_timestamp) return null
  return `people${as_author(itemid)}/${as_type(itemid)}/${closest_timestamp}/${created}`
}
