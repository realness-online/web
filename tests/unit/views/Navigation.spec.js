import { shallow } from 'vue-test-utils'
import * as firebase from 'firebase/app'
import 'firebase/auth'
import flushPromises from 'flush-promises'
import Navigation from '@/views/Navigation'
import Storage, { person_storage, posts_storage } from '@/persistance/Storage'
import itemid from '@/helpers/itemid'
const six_minutes_ago = Date.now() - (1000 * 60 * 6)
const person = {
  first_name: 'Scott',
  last_name: 'Fryxell',
  id: '/+14151234356'
}
const post = {
  created_at: '2017-12-20T23:01:14.310Z',
  statement: 'I like to move it'
}
describe ('@/views/Navigation.vue', () => {
  let wrapper
  beforeEach(async () => {
    jest.spyOn(itemid, 'load').mockImplementation(() => person)
    jest.spyOn(person_storage, 'as_object').mockImplementation(_ => person)
    sessionStorage.setItem('posts-synced', Date.now())
    const currentUser = {
      phoneNumber: '+16282281824'
    }
    const onAuthStateChanged = jest.fn(state_changed => {
      state_changed(currentUser)
    })
    jest.spyOn(firebase, 'auth').mockImplementation(() => {
      return {
        currentUser,
        onAuthStateChanged
      }
    })
    wrapper = shallow(Navigation)
    wrapper.setData({ version: '1.0.0' })
    await flushPromises()
  })
  afterEach(() => {
    sessionStorage.removeItem('posts-synced')
    wrapper.destroy()
  })
  it ('Renders posts and profile for a person', async () => {
    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.find('[itemprop=posts]')).toBeTruthy()
    expect(wrapper.find('[itemref="profile"]')).toBeTruthy()
  })
  it ('Add a post when post-added is emited', async () => {
    expect(wrapper.vm.days.size).toBe(0)
    wrapper.vm.$emit('post-added', post)
    await flushPromises()
    wrapper.vm.add_post(post)
    expect(wrapper.vm.days.size).toBe(1)
    expect(wrapper.find('ol')).toBeTruthy()
  })
  describe ('navigating the application', () => {
    describe ('handling post events', () => {
      it ('posting:false should render the main navigation', () => {
        expect(wrapper.vm.posting).toBe(false)
        expect(wrapper.element).toMatchSnapshot()
      })
      it ('posting:true should hide main navigation', () => {
        wrapper.setData({ posting: true })
        expect(wrapper.element).toMatchSnapshot()
      })
      it ('post-added event should set has_posts to true', () => {
        expect(wrapper.vm.has_posts).toBe(false)
        wrapper.vm.add_post(post)
        expect(wrapper.vm.has_posts).toBe(true)
      })
    })
    describe ('onBoarding()', () => {
      describe ('signed out', () => {
        it ('textarea is the only navigation element to start', () => {
          const signed_out = jest.fn(state_changed => {
            state_changed(null)
          })
          jest.spyOn(firebase, 'auth').mockImplementationOnce(() => {
            return { onAuthStateChanged: signed_out }
          })
          wrapper = shallow(Navigation)
          expect(wrapper.vm.onboarding['has-posts']).toBeFalsy()
          expect(wrapper.vm.onboarding['signed-in']).toBeFalsy()
          expect(wrapper.vm.onboarding['has-friends']).toBeFalsy()
        })
        it ('Profile button is visible when person has posted', () => {
          jest.spyOn(Storage.prototype, 'as_list').mockImplementation(() => {
            return [post]
          })
          const wrapper = shallow(Navigation)
          expect(wrapper.vm.onboarding['has-posts']).toBe(true)
        })
      })
      describe ('signed in', () => {
        it ('Relations is visible', () => {
          wrapper = shallow(Navigation)
          expect(wrapper.vm.onboarding['signed-in']).toBe(true)
        })
        it ('Feed, Events and posters are visible when person has added a friend', () => {
          jest.spyOn(Storage.prototype, 'as_list').mockImplementation(() => {
            return [post]
          })
          wrapper = shallow(Navigation)
          expect(wrapper.vm.onboarding['has-friends']).toBe(true)
        })
      })
    })
    describe ('#user_name', () => {
      it ('Returns \'You\' by default', () => {
        wrapper.vm.me = {}
        expect(wrapper.vm.user_name).toBe('You')
      })
      it ('Returns the users first name if set', () => {
        expect(wrapper.vm.user_name).toBe('Scott')
      })
    })
  })
})
