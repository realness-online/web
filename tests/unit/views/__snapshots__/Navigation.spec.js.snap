// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@/views/Navigation.vue Renders statements and profile for a person 1`] = `
<section
  class="page"
  id="navigation"
>
  <header>
    <h6
      class="app_version"
    >
      
      1.0.0
    
    </h6>
  </header>
   
  <nav>
    <router-link
      class="black"
      tabindex="-1"
      to="/account"
    >
      
      Scott
    
    </router-link>
     
    <router-link
      class="green"
      tabindex="-1"
      to="/events"
    >
      
      Tonight!
    
    </router-link>
     
    <router-link
      class="green"
      tabindex="-1"
      to="/posters"
    >
      
      Posters
    
    </router-link>
     
    <router-link
      class="blue"
      tabindex="-1"
      to="/feed"
    >
      
      Feed
    
    </router-link>
     
    <router-link
      class="blue"
      tabindex="-1"
      to="/relations"
    >
      
      Relations
    
    </router-link>
     
    <!---->
     
    <!---->
  </nav>
   
  <footer
    hidden="hidden"
  >
    <!---->
  </footer>
</section>
`;

exports[`@/views/Navigation.vue navigating the application handling statement events posting:false should render the main navigation 1`] = `
<section
  class="page"
  id="navigation"
>
  <header>
    <h6
      class="app_version"
    >
      
      1.0.0
    
    </h6>
  </header>
   
  <nav>
    <router-link
      class="black"
      tabindex="-1"
      to="/account"
    >
      
      Scott
    
    </router-link>
     
    <router-link
      class="green"
      tabindex="-1"
      to="/events"
    >
      
      Tonight!
    
    </router-link>
     
    <router-link
      class="green"
      tabindex="-1"
      to="/posters"
    >
      
      Posters
    
    </router-link>
     
    <router-link
      class="blue"
      tabindex="-1"
      to="/feed"
    >
      
      Feed
    
    </router-link>
     
    <router-link
      class="blue"
      tabindex="-1"
      to="/relations"
    >
      
      Relations
    
    </router-link>
     
    <!---->
     
    <!---->
  </nav>
   
  <footer
    hidden="hidden"
  >
    <!---->
  </footer>
</section>
`;

exports[`@/views/Navigation.vue navigating the application handling statement events posting:true should hide main navigation 1`] = `
<section
  class="page posting"
  id="navigation"
>
  <header>
    <h6
      class="app_version"
    >
      
      1.0.0
    
    </h6>
  </header>
   
  <nav>
    <!---->
     
    <!---->
     
    <!---->
     
    <!---->
     
    <!---->
     
    <button
      tabindex="-1"
    >
      
      Done
    
    </button>
     
    <!---->
  </nav>
   
  <footer
    hidden="hidden"
  >
    <!---->
  </footer>
</section>
`;
